src/github.com/matrix-org/dendrite/appservice/routing/user.go

```diff
@@ -0,0 +1,60 @@
+package routing
+
+import (
+	"encoding/json"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/util"
+	"io/ioutil"
+	"net/http"
+)
+
+// URIToUIDResponse represents response to an AppService URI to User Id
+// (/_matrix/app/r0/user?uri={url_encoded_uri}) request
+type URIToUIDResponse struct {
+	UserID string `json:"user_id"`
+}
+
+// URIToUID implements `/_matrix/app/r0/user?uri={url_encoded_uri}`, which
+// enables users to contact App Service users directly by taking an encoded
+// URI and turning it into a Matrix ID on the homeserver.
+// https://matrix.org/docs/spec/application_service/unstable.html#user-ids
+func URIToUID(req *http.Request, cfg config.Dendrite) util.JSONResponse {
+	uri := req.URL.Query().Get("uri")
+	if uri == "" {
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: nil,
+		}
+	}
+	baseReqURL := "http://" + string(cfg.Matrix.ServerName) + "/_matrix/app/unstable/thirdparty/user/"
+	//appServices := cfg.Derived.ApplicationServices
+	for _, appservice := range cfg.Derived.ApplicationServices {
+		// Check all the fields associated with each application service
+		if appservice.IsInterestedInUserID(uri) {
+			// call the application service
+			reqURL := baseReqURL + appservice.ID + "?access_token=" + appservice.HSToken +
+				"&fields=" + uri
+			resp, err := http.Get(reqURL)
+			// take the first successful match and send that back to the user
+			if err == nil {
+				body, _ := ioutil.ReadAll(resp.Body)
+				respMap := map[string]interface{}{}
+				err := json.Unmarshal(body, &respMap)
+				if err != nil {
+					panic(err)
```

Comment:
In general you don't want to panic here but return that the content is not acceptable. I don't know how far this goes but in worst case it can kill whole dendrite.

In this situation it might be best to use [UnmarshalJSONRequest](https://github.com/matrix-org/dendrite/blob/c87fb4d/src/github.com/matrix-org/dendrite/clientapi/httputil/httputil.go#L27) where you can return what that method returns if it is not `nil`.